spec.md

```diff
@@ -920,6 +955,101 @@ message ControllerServiceCapability {
 
 ### Node Service RPC
 
+#### `NodePublishDevice`
+
+A Node Plugin MUST implement this RPC call if it has `PUBLISH_UNPUBLISH_DEVICE` node capability.
+This RPC is called by the CO when a workload that wants to use the specified volume is placed (scheduled) on a node.
+The Plugin SHALL assume that this RPC will be executed on the node where the volume will be used.
+This RPC MUST be called by the CO once per node, per volume.
+If the corresponding Controller Plugin has `PUBLISH_UNPUBLISH_VOLUME` controller capability and the Node Plugin has `PUBLISH_UNPUBLISH_DEVICE`, then the CO MUST guarantee that this RPC is called after `ControllerPublishVolume` is called for the given volume on the given node and returns a success.
+
+This operation MUST be idempotent.
+If this RPC failed, or the CO does not know if it failed or not, it MAY choose to call `NodePublishDevice` again, or choose to call `NodeUnpublishDevice`.
+
+```protobuf
+message NodePublishDeviceRequest {
+  // The API version assumed by the CO. This is a REQUIRED field.
+  Version version = 1;
+
+  // The ID of the volume to publish. This field is REQUIRED.
+  string volume_id = 2;
+
+  // The CO SHALL set this field to the value returned by
+  // `ControllerPublishVolume` if the corresponding Controller Plugin
+  // has `PUBLISH_UNPUBLISH_VOLUME` controller capability, and SHALL be
+  // left unset if the corresponding Controller Plugin does not have
+  // this capability. This is an OPTIONAL field.
+  PublishVolumeInfo publish_volume_info = 3;
+
+  // The path to which the volume will be published. It MUST be an
+  // absolute path in the root filesystem of the process serving this
+  // request. The CO SHALL ensure uniqueness of mount_path per volume.
+  // This is a REQUIRED field.
+  string mount_path = 4;
+
+  // The capability of the volume the CO expects the volume to have.
+  // This is a REQUIRED field.
+  VolumeCapability volume_capability = 5;
+
+  // End user credentials used to authenticate/authorize node publish
+  // request. This field is OPTIONAL.
+  Credentials user_credentials = 6;
```

Comment:
I would argue *against* having Credentials in the  `NodePublishDeviceRequest`. This call is called once per volume per node, not once per volume per workload. Imagine the case where two workloads belonging to two different users where one user has permission and the other does not. Since `NodePublishDeviceRequest` is only called once per volume per node, it should not be checking permissions, and instead the `NodePublishVolume` should check user credentials.

Reply:
Hi @saad-ali,

This part confuses me. To me the credentials are more useful for doing the initial device mount in `NodePubishDevice` than the bind mount in `NodePublishVolume`, but maybe I misunderstand the purpose of the credentials?

Reply:
Think of it this way, a single device can be referenced by multiple workloads, but not all workloads may have equal permissions to access the volume.

The credentials passed should authenticate/authorize a given workload (or it's user) access to that volume. However, since `NodePubishDevice` is only called once per node per volume, it will be triggered for the first workload referencing that volume on that node, but not for subsequent workloads. Therefore, the credentials passed to `NodePubishDevice` can't be workload or user specific credentials, because they won't apply to every workload/user.

To prevent inconsistent usage, I think we should eliminate `Credentials` from `NodePubishDevice` and only have it in the call that is guaranteed to be called for every workload, which is `NodePublishVolume`.

That said, taking a step back, the same argument can be applied to `ControllerPublishVolume`, since that is only called once per volume per plugin as well. We should fix that too.

Alternative, if we really want to keep credentials in every call, we should change the language, in the case of `ControllerPublishVolume` and `NodePubishDevice`, to make it clear that in those two cases the credentials cannot be `End user credentials`. But if we do that, we should clarify what the purpose of the credentials is in those cases.

Reply:
> To prevent inconsistent usage, I think we should eliminate Credentials from NodePubishDevice and only have it in the call that is guaranteed to be called for every workload, which is NodePublishVolume.

I agree w/ @saad-ali on this point. Credentials were intended to be workload specific. Unclear how they fit here.

Reply:
Yep. I just need to rebase them out. I'm working on something else ATM but will remove them in the next hour or so.

Reply:
Any thoughts on this?